Spigot-Server-Patches/0594-Damage-with-any-cause.patch

```diff
@@ -0,0 +1,437 @@
++
+     @Override
+     public double getHealth() {
+         return getParent().getHealth();
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+index 9e9bdf9bc2e5c4d72d811fcb439628d946741324..cb6d47cd5ff2598d2e34d0354aff2bd816bbadbc 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+@@ -50,9 +50,11 @@ import org.bukkit.craftbukkit.CraftServer;
+ import org.bukkit.craftbukkit.CraftWorld;
+ import org.bukkit.craftbukkit.entity.memory.CraftMemoryKey;
+ import org.bukkit.craftbukkit.entity.memory.CraftMemoryMapper;
++import org.bukkit.craftbukkit.event.CraftEventFactory;
+ import org.bukkit.craftbukkit.inventory.CraftEntityEquipment;
+ import org.bukkit.craftbukkit.inventory.CraftItemStack;
+ import org.bukkit.craftbukkit.potion.CraftPotionUtil;
++import io.papermc.paper.util.DamageUtils;
+ import org.bukkit.entity.AbstractArrow;
+ import org.bukkit.entity.DragonFireball;
+ import org.bukkit.entity.Egg;
+@@ -79,6 +81,7 @@ import org.bukkit.entity.TippedArrow;
+ import org.bukkit.entity.Trident;
+ import org.bukkit.entity.WitherSkull;
+ import org.bukkit.entity.memory.MemoryKey;
++import org.bukkit.event.entity.EntityDamageEvent;
+ import org.bukkit.event.entity.EntityPotionEffectEvent;
+ import org.bukkit.event.player.PlayerTeleportEvent;
+ import org.bukkit.inventory.EntityEquipment;
+@@ -91,6 +94,8 @@ import org.bukkit.util.BlockIterator;
+ import org.bukkit.util.RayTraceResult;
+ import org.bukkit.util.Vector;
+ 
++import javax.annotation.Nullable;
++
+ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
+     private CraftEntityEquipment equipment;
+ 
+@@ -321,7 +326,7 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
+ 
+     @Override
+     public void damage(double amount) {
+-        damage(amount, null);
++        damage(amount, (org.bukkit.entity.Entity) null); // Paper - the typecast
+     }
+ 
+     @Override
+@@ -337,6 +342,23 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
+         entity.damageEntity(reason, (float) amount);
+     }
+ 
++    // Paper start
++    @Override
++    public void damage(double amount, EntityDamageEvent.DamageCause damageCause) {
++        DamageSource reason = DamageUtils.damageCauseToDamageSource(damageCause, null);
++        entity.damageEntity(reason, (float) amount);
++    }
++
++    @Override
++    public void damage(double amount, EntityDamageEvent.DamageCause damageCause, org.bukkit.entity.Entity source) {
++        Preconditions.checkNotNull(source, "DamagerEntity cannot be null");
++        net.minecraft.server.Entity handle = ((CraftEntity) source).getHandle();
++        DamageSource reason = DamageUtils.damageCauseToDamageSource(damageCause, handle);
++        CraftEventFactory.entityDamage = handle;
```

Comment:
Just throw this into an `if (source != null)` and make the other method default to this with a null source?

Reply:
No, i dont like this personally. Since some damage causes do require damager in order to be resolved. 
For example Projectile without damager will fallback to generic damage type, it would add to confusion.

Reply:
Having two methods does not solve any confusion when there is absolutely no checks to ensure anything is correct. You use nullables everywhere else; there is no need to do it differently here. Add an if-check and just pass it as usual otherwise.